File: tritium-remote/examples/generic_subscription.rs
use simple_logger::SimpleLogger;
use std::env;

#[tokio::main]
async fn main() {
    // logging controlled by the RUST_LOG environment variable
    SimpleLogger::new().env().init().unwrap();

    let auth_token =
        env::var("TRITIUM_AUTH_TOKEN").expect("TRITIUM_AUTH_TOKEN environment variable not set");

    let mut tritium = tritium_remote::connect(
        "ws://localhost:1234",
        &auth_token,
        Some("tritium-remote example - generic subscription".to_string()),
    )
    .await
    .expect("failed to connect");

    let document: &str = "
        subscription {
            worldEvents {
                timestamp
                eventData
            }
        }
    ";

    log::info!("Document: {document}");

    let mut sub = tritium
        .subscription(&document, None)
        .await
        .expect("subscription failed");

    log::info!("Subscription: {sub:?}");

    loop {
        match sub.results.recv().await {
            Some(r) => {
                let data = r.data;
                log::info!("Subscription data: {data:#?}");
            }
            None => {
                break;
            }
        }
    }
}

File: pure-py-tritium-remote/examples/generic_subscription.py
#!/usr/bin/env python3

import os
import asyncio
import tritium_remote


def on_response(response):
    print("Response:", response)


async def main():
    auth_token = os.environ["TRITIUM_AUTH_TOKEN"]
    host = os.environ.get("TRITIUM_HOST", "localhost")

    print("connecting...")
    tritium = await tritium_remote.connect(
        f"ws://{host}:1234",
        auth_token,
        "Python tritium-remote example - generic subscription (world events)",
    )

    document = """
        subscription {
            worldEvents {
                timestamp
                eventData
            }
        }
    """

    subscription = await tritium.subscribe(document, None, on_response)
    print("Subscription:", subscription)


asyncio.run(main())
